### PR TITLE
filter out unqualified bundles and update run_status to inform user i…

### LINKED ISSUE #1800: [Bundle Manager] Filter out unqualified staged bundles and updated their staged_status in UI
This ticket is to filter out unqualified staged bundles that are requested to run a personal worker which hasn't been discovered in the system, meaning either the worker is connected using a different name or it isn't connected to the system.

### PR DESCRIPTION
**Pull request by candicegjing**
Fixed #1800 
Added a new metadata field called `staged_status` (open to discuss of renaming or reusing `run_status`). 
This ticket will filter out staged bundles that are requested to run on the personal worker which hasn't been connected to CodaLab server yet.
 New frontend will look like this:
Screen Shot 2019-12-02 at 12 53 20 PM

### CODE REVIEW

File: codalab/server/bundle_manager.py [hunk 1]
```diff
@@ -692,6 +693,25 @@ def _get_staged_bundles_to_run(self):
                     bundle,
                     {'state': State.FAILED, 'metadata': {'failure_message': failure_message}},
                 )
+            elif bundle.metadata.request_queue:
+                # For those bundles that were requested to run on a private worker, if the private worker does not exist
+                # temporarily, we dispatch those bundles until the private worker connected to CodaLab server.
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Comment should make it clear that what we're doing here is simply updating the `staged_status` field, not actually carrying out any logic.

File: codalab/server/bundle_manager.py [hunk 2]
```diff
@@ -692,6 +693,25 @@ def _get_staged_bundles_to_run(self):
...
+                request_queue = bundle.metadata.request_queue
+                if request_queue:
+                    tagm = re.match('tag=(.+)', request_queue)
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: We should have a utility function for this (there must be other places where we have this logic).

Comment (percyliang) [reply codalab/server/bundle_manager.py]: This could also return `None`, don't crash on that.

File: codalab/server/bundle_manager.py [hunk 3]
```diff
@@ -692,6 +693,25 @@ def _get_staged_bundles_to_run(self):
...
+                if not ([worker for worker in workers.workers() if worker['tag'] == tagm.group(1)]):
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: use `any(...)` and use more local variables for intermediate quantities to make it more readable.

File: codalab/server/bundle_manager.py [hunk 4]
```diff
@@ -692,6 +693,25 @@ def _get_staged_bundles_to_run(self):
...
+                    # TODO: update bundles with batch operation: implement function batch_update_bundles()
```
Comment (percyliang) [inline codalab/server/bundle_manager.py:755]: Not sure you can do that if the message is different for each bundle

File: codalab/server/bundle_manager.py [hunk 5]
```diff
@@ -692,6 +693,25 @@ def _get_staged_bundles_to_run(self):
...
+                    self._model.update_bundle(
+                        bundle,
+                        {
+                            'metadata': {
+                                'staged_status': 'Bundle {} is requested to run on the private worker {} which has '
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: The uuid does not need to be in the mesage

**Review by percyliang — CHANGES_REQUESTED** [review]
Also, should modify the frontend so `staged_status` shows up there.

File: codalab/server/bundle_manager.py (hunk 4 above)
Comment (candicegjing) [reply codalab/server/bundle_manager.py:755]: Found some information online related to this: [link 1](https://stackoverflow.com/questions/54365873/sqlalchemy-update-multiple-rows-in-one-transaction), [link 2](https://docs.sqlalchemy.org/en/13/_modules/examples/performance/bulk_updates.html), [link 3](https://www.tutorialspoint.com/sqlalchemy/sqlalchemy_orm_updating_objects.htm)

File: codalab/bundles/run_bundle.py [hunk 6]
```diff
@@ -44,6 +44,8 @@ class RunBundle(DerivedBundle):
     METADATA_SPECS.append(MetadataSpec('started', int, 'Time when this bundle started executing.', generated=True, formatting='date'))
     METADATA_SPECS.append(MetadataSpec('last_updated', int, 'Time when information about this bundle was last updated.', generated=True, formatting='date'))
     METADATA_SPECS.append(MetadataSpec('run_status', str, 'Execution status of the bundle.', generated=True))
+    METADATA_SPECS.append(MetadataSpec('staged_status', str, 'Detailed information about the status of the staged bundle.', generated=True))
```
Comment (percyliang) [inline codalab/bundles/run_bundle.py]: "Detailed information" => "Information" (simplify)

File: codalab/server/bundle_manager.py [hunk 7]
```diff
@@ -378,11 +377,10 @@ def _filter_and_sort_workers(workers_list, bundle, bundle_resources):
...
         if request_queue:
-            tagm = re.match('tag=(.+)', request_queue)
-            if tagm:
-                workers_list = [worker for worker in workers_list if worker['tag'] == tagm.group(1)]
-            else:
-                # We don't know how to handle this type of request queue argument
+            workers_list = self._get_matched_workers(request_queue, workers_list)
+            # Technically this shouldn't happen any more as all the bundles that are specified to run on
+            # invalid private workers are filtered out in _get_staged_bundles_to_run()
+            if not workers_list:
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: I'd prefer to be more explicit (`workers_list is None` or `len(workers_list) == 0`)

File: codalab/server/bundle_manager.py [hunk 8]
```diff
@@ -637,20 +635,33 @@ def _check_resource_failure(
                 return global_fail_string % (pretty_print(value), pretty_print(global_max))
         return None
 
-    def _get_staged_bundles_to_run(self):
+    def _get_matched_workers(self, request_queue, workers):
+        """
+        Get all of the workers that match with the name of the requested worker
+        :param request_queue: the requested worker
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Be more precise - `request_queue` can specify a tag that can match workers

File: codalab/server/bundle_manager.py [hunk 9]
```diff
@@ -637,20 +635,33 @@ def _check_resource_failure(
...
+        :param workers: a list of workers
+        :return: a list of matched workers
+        """
+        tagm = re.match('tag=(.+)', request_queue)
+        matched_workers = [worker for worker in workers.workers() if worker['tag'] == tagm.group(1)]
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Add `worker_tag = tagm.group(1)`

File: codalab/server/bundle_manager.py [hunk 10]
```diff
@@ -637,20 +635,33 @@ def _check_resource_failure(
...
+        if not matched_workers:
+            return []
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Seems like this is not necessary when `matched_workers = []` (again, be explicit about types)

File: codalab/server/bundle_manager.py [hunk 11]
```diff
@@ -692,6 +703,24 @@ def _get_staged_bundles_to_run(self):
...
                     {'state': State.FAILED, 'metadata': {'failure_message': failure_message}},
                 )
+            elif bundle.metadata.request_queue:
+                # For those bundles that were requested to run on a private worker, if the private worker does not exist
+                # temporarily, we filter out those bundles so that they won't be dispatched to run on workers.
+                request_queue = bundle.metadata.request_queue
+                matched_workers = self._get_matched_workers(request_queue, workers)
+                if not matched_workers:
+                    # TODO: update bundles with batch operation: implement function batch_update_bundles()
+                    self._model.update_bundle(
```
Comment (percyliang) [inline codalab/server/bundle_manager.py:756]: Seems like this is going to get executed a lot even though the `staged_status` doesn't change?  Not sure what to do about that though...

File: tests/server/bundle_manager_test.py [hunk 12]
```diff
@@ -5,9 +5,24 @@
 from codalab.server.bundle_manager import BundleManager
 from codalab.worker.bundle_state import RunResources
 from codalab.bundles import RunBundle
+from codalab.lib.codalab_manager import CodaLabManager
 
 
 class BundleManagerTest(unittest.TestCase):
+    def setUp(self):
```
Comment (percyliang) [inline tests/server/bundle_manager_test.py:12]: What does this have to do with the main functionality of this PR?

**Review by percyliang — CHANGES_REQUESTED** [review]

File: codalab/server/bundle_manager.py (hunk 11 above)
Comment (candicegjing) [reply codalab/server/bundle_manager.py:756]: That's a good point. We can cache those bundles. They shouldn't take too much space.

File: tests/server/bundle_manager_test.py (hunk 12 above)
Comment (candicegjing) [reply tests/server/bundle_manager_test.py:12]: When I refactored the tag matching logic to `_get_matched_workers` as a separate function, I noticed that it can't be called from `_filter_and_sort_workers` as `_get_matched_workers` is a class function but `_filter_and_sort_workers` is declared as a static function. Then I searched through our codebase, I found out that the static function `_filter_and_sort_workers` is only used by `BundleManagerTest`. If this is only created for unit testing, I think it might not be necessary to keep only this function as a static function within the entire `BundleManager` class. Moreover, it makes more sense to make it as a class function instead of a static function. So I removed the `staticmethod` declaration for `_filter_and_sort_workers` and added missing settings for `BundleManagerTest`.

File: codalab/server/bundle_manager.py (hunk 10 above)
Comment (candicegjing) [reply codalab/server/bundle_manager.py]: This is necessary, as the function will return early so that it won't need to spend time going through the rest of the logic.

File: codalab/server/bundle_manager.py [hunk 13]
```diff
@@ -45,6 +45,8 @@ def __init__(self, codalab_manager):
 
         self._make_uuids_lock = threading.Lock()
         self._make_uuids = set()
+        # A set of uuid of bundles whose requested private worker is not discovered in the system
+        self._bundles_wo_matched_workers = set()
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Write it out: "without"

File: codalab/server/bundle_manager.py (hunk 10 above)
Comment (percyliang) [reply codalab/server/bundle_manager.py]: The rest of the function is just `return matched_workers`, no?

File: codalab/server/bundle_manager.py (hunk 11 above)
Comment (percyliang) [reply codalab/server/bundle_manager.py:756]: I don't think you can cache because a bundle can eventually move out of this set.

**Review by percyliang — CHANGES_REQUESTED** [review]

File: codalab/server/bundle_manager.py (hunk 11 above)
Comment (candicegjing) [reply codalab/server/bundle_manager.py:756]: Added logic to handle this.

File: codalab/server/bundle_manager.py [hunk 14]
```diff
@@ -27,6 +27,8 @@
 # When using the REST api, it is allowed to set Memory to 0 but that means the container has unbounded
 # access to the host machine's memory, which we have decided to not allow
 MINIMUM_REQUEST_MEMORY_BYTES = 4 * 1024 * 1024
+# CodaLab public workers regex on both dev and prod environment
+CODALAB_PUBLIC_WORKER_REGEX = re.compile(r"^vm-clws-(prod|dev)-(worker|gpuworker)-(\d)$")
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: I really don't think we should have this hard-coded in here.

File: codalab/server/bundle_manager.py (hunk 13 above)
Comment (percyliang) [reply codalab/server/bundle_manager.py]: To clarify, call it `_bundles_without_matched_workers`

File: codalab/server/bundle_manager.py [hunk 15]
```diff
@@ -386,11 +388,10 @@ def _filter_and_sort_workers(workers_list, bundle, bundle_resources):
...
         if request_queue:
-            tagm = re.match('tag=(.+)', request_queue)
-            if tagm:
-                workers_list = [worker for worker in workers_list if worker['tag'] == tagm.group(1)]
-            else:
-                # We don't know how to handle this type of request queue argument
+            workers_list = self._get_matched_workers(request_queue, workers_list)
+            # Technically this shouldn't happen any more as all the bundles that are specified to run on
+            # invalid private workers are filtered out in _get_staged_bundles_to_run()
+            if len(workers_list) == 0:
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: This check shouldn't be needed, right?  I don't think there's any performance wins and I'd avoid having too many special cases.

File: codalab/server/bundle_manager.py [hunk 16]
```diff
@@ -714,7 +739,27 @@ def _get_staged_bundles_to_run(self):
...
+                        {
+                            'metadata': {
+                                'staged_status': 'Bundle is requested to run on the private worker {} which has '
+                                'not been connected to the CodaLab server yet.'.format(
+                                    request_queue
+                                )
+                            }
+                        },
+                    )
+                    self._bundles_wo_matched_workers.add(bundle.uuid)
```
Comment (percyliang) [inline codalab/server/bundle_manager.py]: This state is only for efficiency, right?  Did we check that it makes a difference in practice?  Again, try to get rid of state as much as possible.

File: codalab/server/bundle_manager.py [hunk 17]
```diff
@@ -714,7 +739,27 @@ def _get_staged_bundles_to_run(self):
...
                     {'state': State.FAILED, 'metadata': {'failure_message': failure_message}},
                 )
+            elif bundle.metadata.request_queue and len(matched_workers) == 0:
+                # For those bundles that were requested to run on a private worker, if the private worker does not exist
+                # temporarily, we filter out those bundles so that they won't be dispatched to run on workers.
+                if bundle.uuid not in self._bundles_wo_matched_workers:
+                    # TODO: update bundles with batch operation: implement function batch_update_bundles()
+                    self._model.update_bundle(
+                        bundle,
+                        {
```
Comment (percyliang) [inline codalab/server/bundle_manager.py:758]: Where do we actually unset `staged_status`?

**Review by percyliang — COMMENTED** [review]
In general, how was this code tested?  Can you include some description of that in the PR?

File: codalab/server/bundle_manager.py (hunk 16 above)
Comment (candicegjing) [reply codalab/server/bundle_manager.py]: I agree that we shouldn't make special state as much as possible. I think what we are doing here is based on a point brought up earlier which is to avoid updating the database with the exact same information. In reality, the number of bundles that don't have private workers matched could vary from time to time. And this is something we can't control. By timing the `update_bundle` function, one query takes `0.0023462772369384766 seconds` when the database is in the same instance as rest-server/bundle-manager/worker. In production, this number will be larger for sure, as the database is hosted in a separate machine. Technically this will improve the performance for sure. But in terms of how much it will improve the performance overall will fully depend on how many bad bundles submitted to the system.

File: codalab/server/bundle_manager.py (hunk 17 above)
Comment (candicegjing) [reply codalab/server/bundle_manager.py:758]: Good point! Updated

**Comment by candicegjing** [comment]
Rebasing this PR seems to be quite messy (Some changes in other PRs show up here). Closing this PR and creating another one.
